### ISSUE TITLE
vulkan: ggml_vk_flash_attn borrows prealloc_y without updating need_sync or invalidating the reuse cache

### ISSUE THREAD

**Issue by mcoffin** [mcoffin/llama.cpp#1]
> Written by Claude Code on behalf of @mcoffin. Filed on this fork as a tracking
> note, not submitted upstream.

## Summary

`ggml_vk_flash_attn` borrows the shared `prealloc_y` scratch buffer for its mask
optimization, but on exit it does not update any of the bookkeeping that tells the
rest of the Vulkan backend that `prealloc_y` was written. Two independent
consequences follow: a skipped write-after-read barrier, and a stale reuse cache
that can feed flash-attention mask data into an unrelated matrix multiply as its
B operand.

Found while investigating an unrelated NaN fault. **This is not the cause of that
fault** (see "Relationship to the NaN investigation" below), but it looks like a
genuine defect on its own.

Line numbers are against `c4dfaa77b`, file
`ggml/src/ggml-vulkan/ggml-vulkan.cpp`.

## Detail

`ggml_vk_flash_attn` uses `ctx->prealloc_y` as the `mask_opt` buffer:

- `:11160` — `mask_opt_buf = use_mask_opt ? ggml_vk_subbuffer(ctx, ctx->prealloc_y, 0) : q_buf;`
- `:11140-11142` — grows `prealloc_size_y` and reallocates if the mask needs more room
- `:11143-11145` — takes the WAR barrier, correctly, by testing `prealloc_y_need_sync`
- `:11206-11209` — writes it with the `fa_mask_opt` dispatch, then takes its RAW barrier
- `:11258-11262` — binds `mask_opt_buf` into the main FA dispatch as its 7th buffer, i.e. reads it

On exit, at `:11263-11265`:

```cpp
if (use_dequant_kv) {
    ctx->prealloc_x_need_sync = true;
}
```

Only `prealloc_x_need_sync` is set, and only under `use_dequant_kv`.
`prealloc_y_need_sync` is never set, and the reuse-cache triple
`prealloc_y_last_pipeline_used` / `prealloc_y_last_tensor_used` /
`prealloc_y_last_k_padded` is never invalidated, even though the buffer contents
were just overwritten and are being read by the dispatch in flight.

### Consequence 1: skipped write-after-read barrier

The next op to write `prealloc_y` is typically a `ggml_vk_mul_mat_*` src1
conversion. Each of those guards its write by testing `prealloc_y_need_sync`
(`:9501`, `:9514`, `:9781`, `:9793`, `:10484`, `:10513`, `:10700`, `:10713`).
Because flash attention left the flag `false`, the guard passes, the barrier is
skipped, and the conversion may overwrite `prealloc_y` while the flash-attention
dispatch is still reading it.

### Consequence 2: stale reuse cache

This one does not need a race to go wrong.

The `prealloc_y` reuse cache exists so that a matmul can skip re-converting `src1`
when the previous conversion into `prealloc_y` used the same `(pipeline, tensor)`
pair. Flash attention overwrites the buffer contents without invalidating that
cache. A later matmul whose `src1` pointer and conversion pipeline still match the
pre-flash-attention values therefore takes the cache-hit path, skips the
conversion entirely, and consumes whatever flash attention left in `prealloc_y` —
mask bitmask data — as its B operand.

Note the cache is cleared at the top of every `ggml_backend_vk_graph_compute`
(`:17606-17609`) and in `ggml_vk_graph_cleanup` (`:16271-16273`), so this requires
the stale hit to occur within the same graph compute as the flash-attention node.

## Reachability

Gated by `use_mask_opt` at `:11055`:

```cpp
mask && nem1 >= 32 && nem0*nem1 > 32768 && ...
```

So it needs flash attention enabled, a mask present, and a mask large enough —
which points at prompt processing with large ubatches rather than single-token
generation, where `nem1` is small. Not verified by measurement.

## Suggested direction

Untested, and offered only as a starting point rather than a proposed patch:
have `ggml_vk_flash_attn` set `ctx->prealloc_y_need_sync = true` and clear the
`prealloc_y_last_*` cache triple whenever `use_mask_opt` is true, mirroring what
the other `prealloc_y` writers already do at `:9542-9545`, `:10542-10547` and
friends.

## Relationship to the NaN investigation

For future reference, since this was found while chasing a different bug — an
intermittent all-NaN logit vector from Qwen3.6-35B-A3B on RADV. This defect is
**not** the cause of that one:

1. That fault reproduces with `--flash-attn off`, where `ggml_vk_flash_attn` never
   runs at all.
2. A corrupt matmul B operand is shared by every output row, so it predicts broad
   corruption. The measured signature there is exactly 128 contiguous NaN floats
   in one column of one head's SSM state, which is one gated-delta-net workgroup.

## Verification status

Everything under "Detail" was verified by reading the source at the cited lines.
Reachability and the suggested direction are unverified. No runtime reproduction
of this defect has been attempted.
